---
title: "Orthologs annotation"
author: "R. Axel W. Wiberg"
date: "17.04.2023" # Most recent changes made.
output:
  html_document:
    df_print: paged
editor_options: 
  chunk_output_type: console
---

This notebook details the annotation of ortholog groups based on differential gene expression experiments from each species.

# 0. Load libraries and functions
```{r,warning=FALSE,message=FALSE,eval=TRUE,echo=TRUE,results='hide'}
library(here)
library(ggplot2)
library(plyr)
library(reshape)
source(here::here("rscripts","ggplot_theme.R"))
# Source the modified version of dscfAllPairsTest.R
source(here::here("rscripts","dscfAllPairsTest.R"))

se <- function(x) {sqrt(var(x, na.rm = TRUE))/sqrt(length(x[which(!is.na(x))]))}

avh_annotator<-function(og,ogs,sp,sp_ann){
  genes<-unlist(strsplit(ogs[og,paste(sp,"_genes",sep="")],","))
  annotations<-sp_ann$expr[match(genes,sp_ann$gene)]
  return(paste(annotations,collapse=","))
}

positional_annotator<-function(og,ogs,sp,sp_ann){
  genes<-unlist(strsplit(ogs[og,paste(sp,"_genes",sep="")],","))
  annotations<-sp_ann$annotation2[match(genes,sp_ann$gene)]
  return(paste(annotations,collapse=","))
}

```

# 1. Load the Proteinortho/PoFF orthologs
```{r}
orthotab<-read.table(here::here("rdata/orthologs","mac_orths_dups3-cs5-alpha0.7.poff.tsv"),
                     sep="\t",header=TRUE)
colnames(orthotab)<-c("n_spp","n_genes","alg_conn","maccli_genes","machtx_genes","maclig_genes")
orthotab$og<-paste("OG",c(0,seq(1:(nrow(orthotab)-1))),sep="")
orthotab$OGname<-paste("OrthoGroup",c(0,seq(1:(nrow(orthotab)-1))),sep="")
row.names(orthotab)<-orthotab$og
```

# 2. Categorise orthogroups by how many species/genes there are
```{r}
# Count the number of genes
orthotab_genes<-orthotab[,c(4,5,6)]
orthotab_counts<-apply(orthotab_genes,MARGIN = c(1,2),function(x){length(unlist(strsplit(x[1],",")))})
for(col in 1:ncol(orthotab_counts)){
  orthotab_counts[which(orthotab_counts[,col]==0),col]<-NA
}
colnames(orthotab_counts)<-c("Maccli","Machtx","Maclig")
orthotab_counts<-as.data.frame(orthotab_counts)
orthotab_counts$og<-paste("OG",c(0,seq(1:(nrow(orthotab_counts)-1))),sep="")

orthotab_counts$og_type<-vector(length=nrow(orthotab_counts))
# Identify 1-to-1 orthologs
orthotab_counts$og_type[which(apply(orthotab_counts[,c(1,2,3)],
                                    MARGIN=1,function(x){all(x==1)}))]<-"1-to-1"
# Identify orthologs where Machtx and Maccli have 1 copy, but Mlig has 2
orthotab_counts$og_type[which(apply(orthotab_counts[,c(1,2,3)],
                                    MARGIN=1,function(x){x[1]==1 & x[2] == 1 & x[3]==2}))]<-"1-to-MligDup"

# Now add these labels to the full orthologs table
orthotab$og_type<-vector(length=nrow(orthotab))
orthotab$og_type[match(orthotab_counts$og[which(orthotab_counts$og_type=="1-to-1")],orthotab$og)]<-"1-to-1"
orthotab$og_type[match(orthotab_counts$og[which(orthotab_counts$og_type=="1-to-MligDup")],orthotab$og)]<-"1-to-MligDup"
```


# 4. Annotate OGs by positional RNA-Seq annotation
```{r}
Maclig_all_res_data<-read.table(here::here("rdata/positional_rnaseq","Maclig_all_DESeq2Results_reduced.tab"),
                         sep="\t",header=TRUE)
# Rename Ovary 1 + Ovary 2 > Ovary
Maclig_all_res_data$annotation2<-gsub("Ovary 1|Ovary 2","Ovary",Maclig_all_res_data$annotation)
tab<-tapply(Maclig_all_res_data$gene,INDEX=list(Maclig_all_res_data$annotation2),length)
tab
tab/sum(tab)

Maccli_all_res_data<-read.table(
  here::here("rdata/positional_rnaseq","Maccli_all_DESeq2Results_reduced.tab"),
                         sep="\t",header=TRUE)
# Rename Ovary 1 + Ovary 2 > Ovary
Maccli_all_res_data$annotation2<-gsub("Ovary 1|Ovary 2","Ovary",Maccli_all_res_data$annotation)
tapply(Maccli_all_res_data$gene,INDEX=list(Maccli_all_res_data$annotation),length)
tab<-tapply(Maccli_all_res_data$gene,INDEX=list(Maccli_all_res_data$annotation2),length)
tab
tab/sum(tab)

Machtx_all_res_data<-read.table(
  here::here("rdata/positional_rnaseq","Machtx_all_DESeq2Results_reduced.tab"),
                         sep="\t",header=TRUE)
# Rename Ovary 1 + Ovary 2 > Ovary
Machtx_all_res_data$annotation2<-gsub("Ovary 1|Ovary 2","Ovary",Machtx_all_res_data$annotation)
tapply(Machtx_all_res_data$gene,INDEX=list(Machtx_all_res_data$annotation),length)
tab<-tapply(Machtx_all_res_data$gene,INDEX=list(Machtx_all_res_data$annotation2),length)
tab
tab/sum(tab)

orthotab[seq(20,30),"maccli_genes"]
sapply(orthotab$og[seq(20,30)],positional_annotator,ogs=orthotab[seq(20,30),],sp="maccli",sp_ann=Maccli_all_res_data)

# Make positional annotation columns to save annotation info in
orthotab$maccli_pos_ann<-vector(length=nrow(orthotab))
anns<-sapply(orthotab$og,positional_annotator,ogs=orthotab,sp="maccli",sp_ann=Maccli_all_res_data)
orthotab$maccli_pos_ann<-anns

orthotab$maclig_pos_ann<-vector(length=nrow(orthotab))
anns<-sapply(orthotab$og,positional_annotator,ogs=orthotab,sp="maclig",sp_ann=Maclig_all_res_data)
orthotab$maclig_pos_ann<-anns

orthotab$machtx_pos_ann<-vector(length=nrow(orthotab))
anns<-sapply(orthotab$og,positional_annotator,ogs=orthotab,sp="machtx",sp_ann=Machtx_all_res_data)
orthotab$machtx_pos_ann<-anns

# Summarise all unique annotations for the og
orthotab$pos_ann<-apply(orthotab[,c("maccli_pos_ann","maclig_pos_ann","machtx_pos_ann")],
                                          MARGIN = 1,function(x){paste(unique(x),collapse=",")})

# Are the annotation strings the same across species.
# i.e. the same number of genes in each species, the same annotations for these
orthotab$pos_same_ann<-apply(orthotab[,c("maccli_pos_ann","maclig_pos_ann","machtx_pos_ann")],
                    MARGIN = 1,function(x){length(unique(x))==1})

# Is there a single, unique, annotation category for the OG, regardless of the number of genes in each species
orthotab$pos_same_uniq_ann<-apply(orthotab[,c("maccli_pos_ann","maclig_pos_ann","machtx_pos_ann")],
                    MARGIN = 1,function(x){length(unique(unlist(strsplit(paste(x,collapse=","),","))))==1})
```


# 3. Annotate OGs by their adult v. hatchling DE result.
```{r}
# Load A v. H results
machtx_res_data<-read.table(here::here("rdata/AvH_contrasts","Machtx_AvH.tab"),
                     header=TRUE,sep="\t")
maccli_res_data<-read.table(here::here("rdata/AvH_contrasts","Maccli_AvH.tab"),
                     header=TRUE,sep="\t")
maclig_res_data<-read.table(here::here("rdata/AvH_contrasts","Maclig_AvH.tab"),
                     header=TRUE,sep="\t")
all_res_dat<-rbind(machtx_res_data,maccli_res_data,maclig_res_data)
row.names(all_res_dat)<-all_res_dat$gene

anns<-sapply(orthotab$og,avh_annotator,ogs=orthotab,sp="maccli",sp_ann=maccli_res_data)
orthotab$maccli_avh_ann<-anns
anns<-sapply(orthotab$og,avh_annotator,ogs=orthotab,sp="maclig",sp_ann=maclig_res_data)
orthotab$maclig_avh_ann<-anns
anns<-sapply(orthotab$og,avh_annotator,ogs=orthotab,sp="machtx",sp_ann=machtx_res_data)
orthotab$machtx_avh_ann<-anns

# Summarise all unique annotations for the og
orthotab$avh_ann<-apply(orthotab[,c("maccli_avh_ann","maclig_avh_ann","machtx_avh_ann")],
                                          MARGIN = 1,function(x){paste(unique(x),collapse=",")})

# Are the annotation strings the same across species.
# i.e. the same number of genes in each species, the same annotations for these
orthotab$avh_same_ann<-apply(orthotab[,c("maccli_avh_ann","maclig_avh_ann","machtx_avh_ann")],
                    MARGIN = 1,function(x){length(unique(x))==1})

# Is there a single, unique, annotation category for the OG, regardless of the number of genes in each species
orthotab$avh_same_uniq_ann<-apply(orthotab[,c("maccli_avh_ann","maclig_avh_ann","machtx_avh_ann")],
                    MARGIN = 1,function(x){length(unique(unlist(strsplit(paste(x,collapse=","),","))))==1})
```

# 4. Save this table
```{r}
head(orthotab)
head(orthotab[which(orthotab$og_type=="1-to-MligDup"),])
head(orthotab[which(orthotab$og_type=="1-to-1"),])
              
write.table(orthotab,
            here::here("rdata/orthologs","mac_orths_dups3-cs5-alpha0.7.poff.pos.avh.tsv"),
            quote=FALSE,row.names=FALSE,col.names=TRUE,sep="\t")
```



