---
title: "Positional RNA-seq analysis"
author: "R. Axel W. Wiberg"
date: "14.04.2023" # Most recent changes made.
output:
  html_document:
    df_print: paged
editor_options: 
  chunk_output_type: console
---

# 0. Load packages
```{r}
library(here)
library(ggplot2)
library(reshape)
library(DESeq2)
library(rgl)
source(here::here("rscripts","ggplot_theme.R"))
se <- function(x) {sqrt(var(x, na.rm = TRUE))/sqrt(length(x))}
```

# 1 Analysis for M. lignano
# 1.1 Load expression data
Read in the data
```{r}
count_dat<-read.table(here::here("rdata/positional_rnaseq","Maclig_raw_counts.csv"),
                      header=TRUE,sep="\t")
tpm_dat<-read.table(here::here("rdata/positional_rnaseq","Maclig_tpm.csv"),
                      header=TRUE,sep="\t")
```

# 1.2. Differential expression analysis with DESeq2
Make a PCA of the raw expression data. Use the TPM values.
```{r}
# For PCA need to also get rid of transcripts variance of 0 across all samples,
# can happen even if there is some counts...
head(tpm_dat)
ncol(tpm_dat)
non_zero_var_tpm<-which(apply(tpm_dat,MARGIN = 1,function(x){var(x)>0}))
length(non_zero_var_tpm)

tpm_dat_filtvar<-tpm_dat[
  which(row.names(tpm_dat) %in% names(non_zero_var_tpm)),]

pca.data<-prcomp(t(tpm_dat_filtvar),center = TRUE,scale. = TRUE)
pca.data$x


pca.plot.data<-as.data.frame(pca.data$x)
pca.plot.data$sample<-rownames(pca.plot.data)
frag_rep<-do.call(rbind,strsplit(pca.plot.data$sample,split = "_"))
colnames(frag_rep)<-c("fragment","rep")
pca.plot.data<-cbind(pca.plot.data,frag_rep)
pca.plot.data

ggplot()+
  geom_label(data=pca.plot.data,
             aes(x=PC1,y=PC2,label=sample,fill=fragment))+
  ggtitle("M. lignano")+
  theme(text=element_text(size=16))
```

First we filter the data a little bit: 
 -We keep only transcripts where at least one sample has count data.
(i.e. we remove all genes that have 0 counts in all samples)
 -We keep only transcripts with a total count > 16
(i.e. average count > 1)
```{r} 
# Need to make a matrix of just the counts
big_all_quant_count<-count_dat[,1:ncol(count_dat)]

# Remove 1st H replicate and 1st T replicate
# This was necessary because we noticed issues with these data.
# These issues are not really apparent in the PCA above (samples cluster by fragment). 
# In several cases (mostly for low expressed genes), H_1 and T_1 had inexplicably similar expression to "whole worm" (W) fragments. 
# It's possible that human error during cutting resulted in contamination.
# e.g. a whole worm, or the wrong part of a cut worm, ended up in the sample.
# See for example: big_all_quant_count["Mlig018904.g1",]

big_all_quant_count<-big_all_quant_count[,c(2,3,4,
                                            6,7,8,
                                            9,10,11,12,
                                            13,14,15,16)]
total<-nrow(big_all_quant_count)
total

# Which transcripts have an counts > 0 in more than 1 sample?
sum_count0_transcripts<-row.names(count_dat[which(apply(count_dat[,2:ncol(count_dat)],1,function(x){sum(x>0)>1})),])
total-length(sum_count0_transcripts)

# Subset the data to include only these transcripts
count_dat<-count_dat[which(row.names(count_dat) %in% sum_count0_transcripts),]
big_all_quant_count<-big_all_quant_count[which(rownames(big_all_quant_count) %in% sum_count0_transcripts),]

# Which transcripts have overall sum of counts >= 16?
sum_count16_transcripts<-row.names(count_dat[which(apply(count_dat[,2:ncol(count_dat)],1,function(x){sum(x)>=16})),])
length(sum_count16_transcripts)
total-length(sum_count16_transcripts)

# Subset the data to include only these transcripts
count_dat<-count_dat[which(row.names(count_dat) %in% sum_count16_transcripts),]
big_all_quant_count<-big_all_quant_count[which(row.names(big_all_quant_count) %in% sum_count16_transcripts),]

# How many transcripts are left?
nrow(big_all_quant_count)
nrow(count_dat)
```

Then we collect the treatment information from the column names of the count matrix and store it separately.
We will need this later.
```{r}
condition<-gsub("_\\d+","",colnames(big_all_quant_count))
coldata<-data.frame("condition"=condition)
```

Then we re-name the column names to have just numbers, so they are the same as the rownames of the "coldata" table.
```{r}
colnames(big_all_quant_count)<-seq(1,ncol(big_all_quant_count))
```

Make sure the sample information maps to the count data. 
i.e. that row names of coldata are all represented among the column names of the count matrix
```{r}
rownames(coldata)<-colnames(big_all_quant_count)
all(rownames(coldata) %in% colnames(big_all_quant_count))
```

Rearrange the count data so that it matches the sample information data. 
This is not strictly necessary here because they way I have loaded the data ensures the correct order. 
But it is important that the columns of the count matrix are in the correct order with respect to the rows in coldata
```{r}
big_all_quant_count <- big_all_quant_count[,rownames(coldata)]
all(rownames(coldata) == colnames(big_all_quant_count))
```

Now we make a DESeq formatted dataset from the count data and the treatment information.
We also need to give the model design.
In this case we want to estimate expression as a function of "condition" (i.e. treatment)
```{r}
dds<-DESeqDataSetFromMatrix(countData=big_all_quant_count,colData=coldata,design= ~ condition)
dds<-estimateSizeFactors(dds)
sizeFactors(dds)
```

We also want to save the normalised counts for future reference.
```{r}
# Compute the normalised counts
normalised_counts<-counts(dds,normalized=TRUE)
# Save the normalised counts as a table
normalised_counts<-as.data.frame(normalised_counts)
colnames(normalised_counts)<-paste(coldata$condition,"_",seq(1,4),sep="")
normalised_counts$gene<-row.names(normalised_counts)

write.table(normalised_counts,
            here::here("rdata/positional_rnaseq","Maclig_normalised_counts.tab"),
            col.names = TRUE,row.names = FALSE,sep="\t",quote=FALSE)
```

Now we run DESeq2
```{r}
dds<-estimateDispersions(dds)
dds
dds<-DESeq(dds)
plotDispEsts(dds)
```

Now we run through 3 contrasts:

Head vs. Testis
Testis vs. Ovary
Ovary vs. Whole worm

For each we take significant DE at an FDR threshhold of 0.05.
```{r}
all_res_data<-data.frame("gene"=row.names(big_all_quant_count))
contrasts<-data.frame("contrast"=c("H_v_T","T_v_O","O_v_W"),
                      "1"=c("H","T","O"),
                      "2"=c("T","O","W"))
contrasts$up<-vector(length=nrow(contrasts))
contrasts$down<-vector(length=nrow(contrasts))
contrasts$outliers<-vector(length=nrow(contrasts))

for(contrast in 1:nrow(contrasts)){
 c<-contrasts$contrast[contrast]
 res<-results(dds,alpha=0.05,
              contrast=c("condition",contrasts$X2[contrast],contrasts$X1[contrast]),
              independentFiltering = FALSE)
  
  # contrast = c("condition",1,2)
  # "up" = lFC > 0 = higher in 1
  # "down" = lFC < 0 = higher in 2
  
  sum_res<-summary(res)
  contrasts$up[contrast]<-nrow(res[which(res$log2FoldChange>0 & res$padj < 0.05),])
  contrasts$down[contrast]<-nrow(res[which(res$log2FoldChange<0 & res$padj < 0.05),])
  contrasts$outliers[contrast]<-nrow(res[which(is.na(res$padj)),])
  
  res_data<-as.data.frame(res)
  res_data$gene<-row.names(res_data)

  write.table(res_data,
            here::here("rdata/positional_rnaseq",paste("Maclig_",c,"_DESeq2Results_reduced.tab",sep="")),
            col.names = TRUE,row.names = FALSE,sep="\t",quote=FALSE)

  all_res_data[[paste(c,"_lFC",sep="")]]<-res_data$log2FoldChange[match(all_res_data$gene,res_data$gene)]
  all_res_data[[paste(c,"_padj",sep="")]]<-res_data$padj[match(all_res_data$gene,res_data$gene)]
  all_res_data[[paste(c,"_expr",sep="")]]<-ifelse(
    (all_res_data[[paste(c,"_lFC",sep="")]] > 0 & all_res_data[[paste(c,"_padj",sep="")]] < 0.05),"+",ifelse(
      (all_res_data[[paste(c,"_lFC",sep="")]] < 0 & all_res_data[[paste(c,"_padj",sep="")]] < 0.05),"-","0"))
}
contrasts
all_res_data$pat<-paste(all_res_data$H_v_T_expr,all_res_data$T_v_O_expr,all_res_data$O_v_W_expr,sep=",")
```

Annotate each gene by it's pattern of gene expression across fragments.
```{r}
tapply(all_res_data$gene,INDEX=list(all_res_data$pat),length)
all_res_data$annotation<-ifelse(all_res_data$pat=="+,0,0","Testis",
                                ifelse(all_res_data$pat=="0,+,0","Ovary 1",
                                       ifelse(all_res_data$pat=="+,+,0","Ovary 2",
                                              ifelse(all_res_data$pat=="0,0,+","Tail",
                                                     ifelse(all_res_data$pat=="0,0,0","Non-Specific","Other")))))
# Save the results
write.table(all_res_data,
            here::here("rdata/positional_rnaseq","Maclig_all_DESeq2Results_reduced.tab"),
            col.names = TRUE,row.names = FALSE,sep="\t",quote=FALSE)
```

# 1.3 Make a 3D plot as in previous studies
```{r}
# Subset count dat and tpm dat to those that received annotations
count_dat<-count_dat[which(row.names(count_dat) %in% all_res_data$gene),]
tpm_dat<-tpm_dat[which(row.names(tpm_dat) %in% all_res_data$gene),]
normalised_counts<-normalised_counts[which(normalised_counts$gene %in% all_res_data$gene),]

# Compute the mean TPM from across replicates of each fragment
mean_count_dat<-data.frame("gene"=normalised_counts$gene)

mean_count_dat$W_mean<-apply(normalised_counts[,grep("W_",colnames(normalised_counts))],MARGIN = 1,mean)
mean_count_dat$W_expr<-log2(((mean_count_dat$W_mean/sum(mean_count_dat$W_mean)*100)+0.00001))

mean_count_dat$H_mean<-apply(normalised_counts[,grep("H_",colnames(normalised_counts))],MARGIN = 1,mean)
mean_count_dat$H_expr<-log2(((mean_count_dat$H_mean/sum(mean_count_dat$H_mean)*100)+0.00001))

mean_count_dat$T_mean<-apply(normalised_counts[,grep("T_",colnames(normalised_counts))],MARGIN = 1,mean)
mean_count_dat$T_expr<-log2(((mean_count_dat$T_mean/sum(mean_count_dat$T_mean)*100)+0.00001))

mean_count_dat$O_mean<-apply(normalised_counts[,grep("O_",colnames(normalised_counts))],MARGIN = 1,mean)
mean_count_dat$O_expr<-log2(((mean_count_dat$O_mean/sum(mean_count_dat$O_mean)*100)+0.00001))

mean_count_dat$overall_mean<-apply(count_dat[,grep("_",colnames(count_dat))],MARGIN = 1,mean)

head(mean_count_dat)
mean_count_dat$T_H<-mean_count_dat$T_expr-mean_count_dat$H_expr
mean_count_dat$O_T<-mean_count_dat$O_expr-mean_count_dat$T_expr
mean_count_dat$W_O<-mean_count_dat$W_expr-mean_count_dat$O_expr


mean_count_dat$annotation<-all_res_data$annotation[match(mean_count_dat$gene,all_res_data$gene)]
mean_count_dat$col<-ifelse(mean_count_dat$annotation=="Non-Specific","red",
                           ifelse(mean_count_dat$annotation=="Other","black",
                                  ifelse(mean_count_dat$annotation=="Testis","blue",
                                         ifelse(mean_count_dat$annotation=="Ovary 1","Orange",
                                                ifelse(mean_count_dat$annotation=="Ovary 2","purple",
                                                       ifelse(mean_count_dat$annotation=="Tail","green","grey50"))))))
# Plot 3D scatter plot
setupKnitr()
plot3d(x=mean_count_dat$T_H,
       y=mean_count_dat$O_T,
       z=mean_count_dat$W_O,
       col = mean_count_dat$col,
       xlab = "testis expr - head expr",
       ylab = "ovary expr - testis expr",
       zlab = "whole worm expr - ovary expr",
       type = "p",
       size = 2,
       axis = FALSE
       )
lignano_scatter <- rglwidget(scene3d(minimal = TRUE))
lignano_scatter
```

# 1.4 Compare to annotations from Brand et al. 2020
```{r}
# Read in the results and annotations from this study.
all_res_data<-read.table(here::here("rdata/positional_rnaseq","Maclig_all_DESeq2Results_reduced.tab"),
                         sep="\t",header=TRUE)

# Load the tpm data from salmon for all genes
tpm_dat<-read.table(here::here("rdata/positional_rnaseq","Maclig_tpm.csv"),header=TRUE)
count_dat<-read.table(here::here("rdata/positional_rnaseq","Maclig_raw_counts.csv"),header=TRUE)

# Load the normalised count data from DESeq2 for all genes
normalised_counts<-read.table(here::here("rdata/positional_rnaseq","Maclig_normalised_counts.tab"),sep="\t",header=TRUE)
total<-nrow(count_dat)

# Load the annotations from Brand et al. 2020
brand2020<-read.table(here::here("rdata/positional_rnaseq","Brand_et_al_2020_DACH_Additional_file_14.csv"),
                      header=TRUE,sep="\t")
# Simplify the ID column
brand2020$ID<-gsub("Maclig_37v3@","",brand2020$ID)
# Re-label the old annotations to make consistent with new ones.
brand2020$class_name[brand2020$class_name=="Testes"]<-"Testis"
brand2020$class_name[brand2020$class_name=="Ovaries 1"]<-"Ovary 1"
brand2020$class_name[brand2020$class_name=="Ovaries 2"]<-"Ovary 2"

# Add the Brand et al. annotation to the new dataset
all_res_data$brand2020<-brand2020$class_name[match(all_res_data$gene,brand2020$ID)]

# Do the annotations agree?
all_res_data$agree<-all_res_data$annotation==all_res_data$brand2020

#Rows = this study, columns = Brand et al. 2020
head(all_res_data)
table(all_res_data$annotation,all_res_data$brand2020)

# How many agree, how many don't?
table(all_res_data$agree,all_res_data$annotation)
sum(table(all_res_data$agree,all_res_data$annotation))
```

# 1.5 Compare individual expression levels across studies.
```{r}
# Add the expression estimates from Brand et al. 2020 for each fragment
mean_count_dat$brand2020_W<-brand2020$D_expr[match(mean_count_dat$gene,brand2020$ID)]
mean_count_dat$brand2020_H<-brand2020$A_expr[match(mean_count_dat$gene,brand2020$ID)]
mean_count_dat$brand2020_T<-brand2020$B_expr[match(mean_count_dat$gene,brand2020$ID)]
mean_count_dat$brand2020_O<-brand2020$C_expr[match(mean_count_dat$gene,brand2020$ID)]

mean_count_dat$brand2020_T_H<-mean_count_dat$brand2020_T-mean_count_dat$brand2020_H
mean_count_dat$brand2020_O_T<-mean_count_dat$brand2020_O-mean_count_dat$brand2020_T
mean_count_dat$brand2020_W_O<-mean_count_dat$brand2020_W-mean_count_dat$brand2020_O

mean_count_dat$brand2020_annotation<-all_res_data$brand2020[match(mean_count_dat$gene,all_res_data$gene)]
mean_count_dat$brand2020_col<-ifelse(
  mean_count_dat$brand2020_annotation=="Non-Specific","red",ifelse(
    mean_count_dat$brand2020_annotation=="Other","black",ifelse(
      mean_count_dat$brand2020_annotation=="Testis","blue",ifelse(
        mean_count_dat$brand2020_annotation=="Ovary 1","Orange",ifelse(
          mean_count_dat$brand2020_annotation=="Ovary 2","purple",ifelse(
            mean_count_dat$brand2020_annotation=="Tail","green","grey50"))))))

mean_count_dat$agree<-all_res_data$agree[match(mean_count_dat$gene,all_res_data$gene)]

for(frag in c("H","T","O","W")){
  cat(frag,"\n")
  plot_dat<-mean_count_dat[,c("gene","agree",paste(frag,"_mean",sep=""),paste("brand2020_",frag,sep=""))]
  colnames(plot_dat)<-c("gene","agree","mean","brand2020_mean")

  test<-cor.test(plot_dat$mean[which(plot_dat$agree==TRUE)],
                 plot_dat$brand2020_mean[which(plot_dat$agree==TRUE)],method = "spearman")
  cat("AGREE: rho = ",test$estimate,"- S = ",test$statistic,"- p =",test$p.value,"\n")
  test<-cor.test(plot_dat$mean[which(plot_dat$agree==FALSE)],
                 plot_dat$brand2020_mean[which(plot_dat$agree==FALSE)],method = "spearman")
  cat("NOT AGREE: rho = ",test$estimate,"- S = ",test$statistic,"- p =",test$p.value,"\n")

  head(plot_dat)
  plot<-ggplot()+
    geom_point(data=plot_dat,
               aes(log(mean),brand2020_mean,colour=agree))+
    ylab("Brand 2020 expression")+
    xlab("Mean TPM")+
    ggtitle(frag)+
    facet_grid(agree~.)
  print(plot)
}
```

# 1.6 Compare annotations from bioinformatics to annotations from ISH
Check annotations based on ISH patterns against annotations based on positional RNA-seq in M. lignano
I have assigned old transcript IDs (from which ISH primers were developed) to new Mlig transcripts.
I have also assigned annotations from Arbore et al. 2015, Brand et al. 2020 and the current study to these transcripts.
```{r}
# Load the primers info table
ish_primers_ann<-read.table(here::here("rdata/positional_rnaseq/ish","primer_sets_full.csv"),header=TRUE,sep="\t")
nrow(ish_primers_ann)
# Out of the Arbore et al. 2015 "Tail" annotations, how many are also "Tail" annotated from ISH patterns
ish_primers_ann_tail<-ish_primers_ann[which(ish_primers_ann$ann_Arbore2015=="Tail"),]
ish_primers_ann_tail$same<-ish_primers_ann_tail$ann_ISH==ish_primers_ann_tail$ann_Arbore2015

cat("N = ",length(ish_primers_ann_tail$same),
    ", N(same) = ",sum(ish_primers_ann_tail$same),
    " (",round(sum(ish_primers_ann_tail$same)/length(ish_primers_ann_tail$same),2),"%)")

#head(ish_primers_ann)
# Out of the Brand et al. 2020  "Tail" annotations, how many are also "Tail" annotated from ISH patterns
ish_primers_ann_tail<-ish_primers_ann[which(ish_primers_ann$ann_Brand2020=="Tail"),]
ish_primers_ann_tail$same<-ish_primers_ann_tail$ann_ISH==ish_primers_ann_tail$ann_Brand2020

cat("N = ",length(ish_primers_ann_tail$same),
    ", N(same) = ",sum(ish_primers_ann_tail$same),
    " (",round(sum(ish_primers_ann_tail$same)/length(ish_primers_ann_tail$same),2),"%)")
# We are slightly more "precise" in our annotations.
# Although we annotate fewer Tail region transcripts,
# those that we do annotate from RNAseq are more likely to be tail in ISH


# Out of the Wiberg et al.  "Tail" annotations, how many are also "Tail" annotated from ISH patterns
ish_primers_ann_tail<-ish_primers_ann[which(ish_primers_ann$ann_Wiberg=="Tail"),]
ish_primers_ann_tail$same<-ish_primers_ann_tail$ann_ISH==ish_primers_ann_tail$ann_Wiberg

cat("N = ",length(ish_primers_ann_tail$same),
    ", N(same) = ",sum(ish_primers_ann_tail$same),
    " (",round(sum(ish_primers_ann_tail$same)/length(ish_primers_ann_tail$same),2),"%)")
# We are slightly more "precise" in our annotations.
# Although we annotate fewer Tail region transcripts,
# those that we do annotate from RNA-Seq are more likely to be tail in ISH
```



# 2. Analysis for M. cliftonense
# 2.1 Load expression data
```{r}
count_dat<-read.table(here::here("rdata/positional_rnaseq","Maccli_raw_counts.csv"),
                      header=TRUE,sep="\t")
tpm_dat<-read.table(here::here("rdata/positional_rnaseq","Maccli_tpm.csv"),
                      header=TRUE,sep="\t")
```

# 2.2. Differential expression analysis with DESeq2
Make a PCA of the raw expression data. Use the TPM values.
```{r}
# For PCA need to also get rid of transcripts variance of 0 across all samples,
# can happen even if there is some counts...
non_zero_var_tpm<-which(apply(tpm_dat,MARGIN = 1,function(x){var(x)>0}))

tpm_dat_filtvar<-tpm_dat[
  which(row.names(tpm_dat) %in% names(non_zero_var_tpm)),]

pca.data<-prcomp(t(tpm_dat_filtvar),center = TRUE,scale. = TRUE)
pca.data$x

pca.plot.data<-as.data.frame(pca.data$x)
pca.plot.data$sample<-rownames(pca.plot.data)
frag_rep<-do.call(rbind,strsplit(pca.plot.data$sample,split = "_"))
colnames(frag_rep)<-c("fragment","rep")
pca.plot.data<-cbind(pca.plot.data,frag_rep)
pca.plot.data

ggplot()+
  geom_label(data=pca.plot.data,aes(x=PC1,y=PC2,label=sample,fill=fragment))+
  ggtitle("M. cliftonense")+
  theme(text=element_text(size=16))
```

First we filter the data a little bit: 
 -We keep only transcripts where at least one sample has count data.
(i.e. we remove all genes that have 0 counts in all samples)
 -We keep only transcripts with a total count > 16
(i.e. average count > 1)
```{r} 
# Need to make a matrix of just the counts
big_all_quant_count<-count_dat[,1:ncol(count_dat)]

total<-nrow(big_all_quant_count)
total

# Which transcripts have an counts > 0 in more than 1 sample?
sum_count0_transcripts<-row.names(count_dat[which(apply(count_dat[,1:ncol(count_dat)],1,function(x){sum(x>0)>1})),])
total-length(sum_count0_transcripts)
# Subset the data to include only these transcripts
count_dat<-count_dat[which(row.names(count_dat) %in% sum_count0_transcripts),]
big_all_quant_count<-big_all_quant_count[which(rownames(big_all_quant_count) %in% sum_count0_transcripts),]

# Which transcripts have overall sum of counts >= 16?
sum_count16_transcripts<-row.names(count_dat[which(apply(count_dat[,1:ncol(count_dat)],1,function(x){sum(x)>=16})),])
length(sum_count16_transcripts)
total-length(sum_count16_transcripts)
# Subset the data to include only these transcripts
count_dat<-count_dat[which(row.names(count_dat) %in% sum_count16_transcripts),]
big_all_quant_count<-big_all_quant_count[which(row.names(big_all_quant_count) %in% sum_count16_transcripts),]
```

Then we collect the treatment information from the column names of the count matrix and store it separately.
We will need this later.
```{r}
condition<-gsub("_\\d+","",colnames(big_all_quant_count))
coldata<-data.frame("condition"=condition)
```

Then we re-name the column names to have just numbers, so they are the same as the rownames of the "coldata" table.
```{r}
colnames(big_all_quant_count)<-seq(1,ncol(big_all_quant_count))
```

Make sure the sample information maps to the count data. 
i.e. that row names of coldata are all represented among the column names of the count matrix
```{r}
rownames(coldata)<-colnames(big_all_quant_count)
all(rownames(coldata) %in% colnames(big_all_quant_count))
```

Rearrange the count data so that it matches the sample information data. 
This is not strictly necessary here because they way I have loaded the data ensures the correct order. 
But it is important that the columns of the count matrix are in the correct order with respect to the rows in coldata
```{r}
big_all_quant_count <- big_all_quant_count[,rownames(coldata)]
all(rownames(coldata) == colnames(big_all_quant_count))
```

Now we make a DESeqw formatted dataset from the count data and the treatment information.
We also need to give the model design. In this case we want to estimate expression as a function of "condition" (i.e. treatment)
```{r}
dds<-DESeqDataSetFromMatrix(countData=big_all_quant_count,colData=coldata,design= ~ condition)
dds<-estimateSizeFactors(dds)
sizeFactors(dds)
```

We also want to save the normalised counts for future reference.
```{r}
# Compute the normalised counts
normalised_counts<-counts(dds,normalized=TRUE)
# Save the normalised counts as a table
normalised_counts<-as.data.frame(normalised_counts)
colnames(normalised_counts)<-paste(coldata$condition,"_",seq(1,4),sep="")
normalised_counts$gene<-row.names(normalised_counts)

write.table(normalised_counts,
            here::here("rdata/positional_rnaseq","Maccli_normalised_counts.tab"),
            col.names = TRUE,row.names = FALSE,sep="\t",quote=FALSE)
```

Now we run DESeq2
```{r}
dds<-estimateDispersions(dds)
dds
dds<-DESeq(dds)
plotDispEsts(dds)
```

Now we run through 3 contrasts:

Head vs. Testis
Testis vs. Ovary
Ovary vs. Whole worm

For each we take significant DE at an FDR threshhold of 0.05.
```{r}
all_res_data<-data.frame("gene"=row.names(big_all_quant_count))
contrasts<-data.frame("contrast"=c("H_v_T","T_v_O","O_v_W"),
                      "1"=c("H","T","O"),
                      "2"=c("T","O","W"))
contrasts$up<-vector(length=nrow(contrasts))
contrasts$down<-vector(length=nrow(contrasts))
contrasts$outliers<-vector(length=nrow(contrasts))

for(contrast in 1:nrow(contrasts)){
 c<-contrasts$contrast[contrast]
 res<-results(dds,alpha=0.05,
              contrast=c("condition",contrasts$X2[contrast],contrasts$X1[contrast]),
              independentFiltering = FALSE)
  # contrast = c("condition",1,2)
  # "up" = lFC > 0 = higher in 1
  # "down" = lFC < 0 = higher in 2
  sum_res<-summary(res)
  contrasts$up[contrast]<-nrow(res[which(res$log2FoldChange>0 & res$padj < 0.05),])
  contrasts$down[contrast]<-nrow(res[which(res$log2FoldChange<0 & res$padj < 0.05),])
  contrasts$outliers[contrast]<-nrow(res[which(is.na(res$padj)),])
  
  res_data<-as.data.frame(res)
  res_data$gene<-row.names(res_data)

  write.table(res_data,
            here::here("rdata/positional_rnaseq",paste("Maccli_",c,"_DESeq2Results_reduced.tab",sep="")),
            col.names = TRUE,row.names = FALSE,sep="\t",quote=FALSE)

  all_res_data[[paste(c,"_lFC",sep="")]]<-res_data$log2FoldChange[match(all_res_data$gene,res_data$gene)]
  all_res_data[[paste(c,"_padj",sep="")]]<-res_data$padj[match(all_res_data$gene,res_data$gene)]
  all_res_data[[paste(c,"_expr",sep="")]]<-ifelse(
    (all_res_data[[paste(c,"_lFC",sep="")]] > 0 & all_res_data[[paste(c,"_padj",sep="")]] < 0.05),"+",ifelse(
      (all_res_data[[paste(c,"_lFC",sep="")]] < 0 & all_res_data[[paste(c,"_padj",sep="")]] < 0.05),"-","0"))
}
contrasts
all_res_data$pat<-paste(all_res_data$H_v_T_expr,all_res_data$T_v_O_expr,all_res_data$O_v_W_expr,sep=",")
```

Annotate each gene by it's pattern of gene expression across fragments.
```{r}
tapply(all_res_data$gene,INDEX=list(all_res_data$pat),length)
all_res_data$annotation<-ifelse(all_res_data$pat=="+,0,0","Testis",
                                ifelse(all_res_data$pat=="0,+,0","Ovary 1",
                                       ifelse(all_res_data$pat=="+,+,0","Ovary 2",
                                              ifelse(all_res_data$pat=="0,0,+","Tail",
                                                     ifelse(all_res_data$pat=="0,0,0","Non-Specific","Other")))))
head(all_res_data)
head(all_res_data[which(all_res_data$pat=="+,0,0"),])
tapply(all_res_data$gene,INDEX=list(all_res_data$annotation),length)

# Save the results
write.table(all_res_data,
            here::here("rdata/positional_rnaseq","Maccli_all_DESeq2Results_reduced.tab"),
            col.names = TRUE,row.names = FALSE,sep="\t",quote=FALSE)
```

Plot 3D scatter
```{r}
all_res_data<-read.table(here::here("rdata/positional_rnaseq","Maccli_all_DESeq2Results_reduced.tab"),
                         header=TRUE,sep="\t")
normalised_counts<-read.table(here::here("rdata/positional_rnaseq","Maccli_normalised_counts.tab"),sep="\t",header=TRUE)
total<-nrow(count_dat)

# Subset count dat and tpm dat to those that received annotations
normalised_counts<-normalised_counts[which(normalised_counts$gene %in% all_res_data$gene),]
nrow(normalised_counts)

# Compute the mean TPM from across replicates of each fragment
mean_count_dat<-data.frame("gene"=normalised_counts$gene)
mean_count_dat$W_mean<-apply(normalised_counts[,grep("W_",colnames(normalised_counts))],MARGIN = 1,mean)
mean_count_dat$W_expr<-log2(((mean_count_dat$W_mean/sum(mean_count_dat$W_mean)*100)+0.00001))
mean_count_dat$H_mean<-apply(normalised_counts[,grep("H_",colnames(normalised_counts))],MARGIN = 1,mean)
mean_count_dat$H_expr<-log2(((mean_count_dat$H_mean/sum(mean_count_dat$H_mean)*100)+0.00001))
mean_count_dat$T_mean<-apply(normalised_counts[,grep("T_",colnames(normalised_counts))],MARGIN = 1,mean)
mean_count_dat$T_expr<-log2(((mean_count_dat$T_mean/sum(mean_count_dat$T_mean)*100)+0.00001))
mean_count_dat$O_mean<-apply(normalised_counts[,grep("O_",colnames(normalised_counts))],MARGIN = 1,mean)
mean_count_dat$O_expr<-log2(((mean_count_dat$O_mean/sum(mean_count_dat$O_mean)*100)+0.00001))
mean_count_dat$overall_mean<-apply(count_dat[,grep("_",colnames(count_dat))],MARGIN = 1,mean)

mean_count_dat$T_H<-mean_count_dat$T_expr-mean_count_dat$H_expr
mean_count_dat$O_T<-mean_count_dat$O_expr-mean_count_dat$T_expr
mean_count_dat$W_O<-mean_count_dat$W_expr-mean_count_dat$O_expr

mean_count_dat$annotation<-all_res_data$annotation[match(mean_count_dat$gene,all_res_data$gene)]
mean_count_dat$col<-ifelse(mean_count_dat$annotation=="Non-Specific","red",
                           ifelse(mean_count_dat$annotation=="Other","black",
                                  ifelse(mean_count_dat$annotation=="Testis","blue",
                                         ifelse(mean_count_dat$annotation=="Ovary 1","Orange",
                                                ifelse(mean_count_dat$annotation=="Ovary 2","purple",
                                                       ifelse(mean_count_dat$annotation=="Tail","green","grey50"))))))

head(mean_count_dat)

# Plot 3D scatter plot
plot3d(x=mean_count_dat$T_H,
       y=mean_count_dat$O_T,
       z=mean_count_dat$W_O,
       col = mean_count_dat$col,
       xlab = "testis expr - head expr",
       ylab = "ovary expr - testis expr",
       zlab = "whole worm expr - ovary expr",
       type = "p",
       size = 2,
       axis = FALSE
       )
cliftonense_scatter <- rglwidget(scene3d(minimal = TRUE))
cliftonense_scatter
```



# 3 Analysis for M. hystrix
# 3.1 Load expression data
```{r}
count_dat<-read.table(here::here("rdata/positional_rnaseq","Machtx_raw_counts.csv"),
                      header=TRUE,sep="\t")
tpm_dat<-read.table(here::here("rdata/positional_rnaseq","Machtx_tpm.csv"),
                      header=TRUE,sep="\t")
```

# 3.2. Differential expression analysis with DESeq2
Make a PCA of the raw expression data. Use the TPM values.
```{r}
# For PCA need to also get rid of transcripts variance of 0 across all samples,
# can happen even if there is some counts...
non_zero_var_tpm<-which(apply(tpm_dat,MARGIN = 1,function(x){var(x)>0}))

tpm_dat_filtvar<-tpm_dat[
  which(row.names(tpm_dat) %in% names(non_zero_var_tpm)),]

pca.data<-prcomp(t(tpm_dat_filtvar),center = TRUE,scale. = TRUE)
pca.data$x


pca.plot.data<-as.data.frame(pca.data$x)
pca.plot.data$sample<-rownames(pca.plot.data)
frag_rep<-do.call(rbind,strsplit(pca.plot.data$sample,split = "_"))
colnames(frag_rep)<-c("fragment","rep")
pca.plot.data<-cbind(pca.plot.data,frag_rep)
pca.plot.data

ggplot()+
  geom_label(data=pca.plot.data,aes(x=PC1,y=PC2,label=sample,fill=fragment))+
  ggtitle("M. hystrix")+
  theme(text=element_text(size=16))
```


First we filter the data a little bit: 
 -We keep only transcripts where at least one sample has count data.
(i.e. we remove all genes that have 0 counts in all samples)
 -We keep only transcripts with a total count > 16
(i.e. average count > 1)
```{r} 
# Need to make a matrix of just the counts
big_all_quant_count<-count_dat[,1:ncol(count_dat)]

total<-nrow(big_all_quant_count)
total

# Which transcripts have an counts > 0 in more than 1 sample
sum_count0_transcripts<-row.names(count_dat[which(apply(count_dat[,1:ncol(count_dat)],1,function(x){sum(x>0)>1})),])
total-length(sum_count0_transcripts)
# Subset the data to include only these transcripts
count_dat<-count_dat[which(row.names(count_dat) %in% sum_count0_transcripts),]
big_all_quant_count<-big_all_quant_count[which(rownames(big_all_quant_count) %in% sum_count0_transcripts),]


# Which transcripts have overall sum of counts >= 16
sum_count16_transcripts<-row.names(count_dat[which(apply(count_dat[,1:ncol(count_dat)],1,function(x){sum(x)>=16})),])
length(sum_count16_transcripts)
total-length(sum_count16_transcripts)
# Subset the data to include only these transcripts
count_dat<-count_dat[which(row.names(count_dat) %in% sum_count16_transcripts),]
big_all_quant_count<-big_all_quant_count[which(row.names(big_all_quant_count) %in% sum_count16_transcripts),]

```

Then we collect the treatment information from the column names of the count matrix and store it separately.
We will need this later.
```{r}
condition<-gsub("_\\d+","",colnames(big_all_quant_count))
coldata<-data.frame("condition"=condition)
head(big_all_quant_count,2)
```

Then we re-name the column names to have just numbers, so they are the same as the rownames of the "coldata" table.
```{r}
colnames(big_all_quant_count)<-seq(1,ncol(big_all_quant_count))
```

Make sure the sample information maps to the count data. 
i.e. that row names of coldata are all represented among the column names of the count matrix
```{r}
rownames(coldata)<-colnames(big_all_quant_count)
all(rownames(coldata) %in% colnames(big_all_quant_count))
```

Rearrange the count data so that it matches the sample information data. 
This is not strictly necessary here because they way I have loaded the data ensures the correct order. 
But it is important that the columns of the count matrix are in the correct order with respect to the rows in coldata
```{r}
big_all_quant_count <- big_all_quant_count[,rownames(coldata)]
all(rownames(coldata) == colnames(big_all_quant_count))
```

Now we make a DESeqw formatted dataset from the count data and the treatment information.
We also need to give the model design. In this case we want to estimate expression as a function of "condition" (i.e. treatment)
```{r}
dds<-DESeqDataSetFromMatrix(countData=big_all_quant_count,colData=coldata,design= ~ condition)
dds<-estimateSizeFactors(dds)
sizeFactors(dds)
```

We also want to save the normalised counts for future reference.
```{r}
# Compute the normalised counts
normalised_counts<-counts(dds,normalized=TRUE)
# Save the normalised counts as a table
normalised_counts<-as.data.frame(normalised_counts)
colnames(normalised_counts)<-paste(coldata$condition,"_",seq(1,4),sep="")
normalised_counts$gene<-row.names(normalised_counts)

write.table(normalised_counts,
            here::here("rdata/positional_rnaseq","Machtx_normalised_counts.tab"),
            col.names = TRUE,row.names = FALSE,sep="\t",quote=FALSE)
```


Now we run DESeq2
```{r}
dds<-estimateDispersions(dds)
dds
dds<-DESeq(dds)
plotDispEsts(dds)
```

Now we run through 3 contrasts:

Head vs. Testis
Testis vs. Ovary
Ovary vs. Whole worm

For each we take significant DE at an FDR threshhold of 0.05.
```{r}
all_res_data<-data.frame("gene"=row.names(big_all_quant_count))
contrasts<-data.frame("contrast"=c("H_v_T","T_v_O","O_v_W"),
                      "1"=c("H","T","O"),
                      "2"=c("T","O","W"))
contrasts$up<-vector(length=nrow(contrasts))
contrasts$down<-vector(length=nrow(contrasts))
contrasts$outliers<-vector(length=nrow(contrasts))

for(contrast in 1:nrow(contrasts)){
 c<-contrasts$contrast[contrast]
 res<-results(dds,alpha=0.05,
              contrast=c("condition",contrasts$X2[contrast],contrasts$X1[contrast]),
              independentFiltering = FALSE)
  # contrast = c("condition",1,2)
  # "up" = lFC > 0 = higher in 1
  # "down" = lFC < 0 = higher in 2
  sum_res<-summary(res)
  contrasts$up[contrast]<-nrow(res[which(res$log2FoldChange>0 & res$padj < 0.05),])
  contrasts$down[contrast]<-nrow(res[which(res$log2FoldChange<0 & res$padj < 0.05),])
  contrasts$outliers[contrast]<-nrow(res[which(is.na(res$padj)),])
  
  res_data<-as.data.frame(res)
  res_data$gene<-row.names(res_data)

  write.table(res_data,
            here::here("rdata/positional_rnaseq",paste("Machtx_",c,"_DESeq2Results_reduced.tab",sep="")),
            col.names = TRUE,row.names = FALSE,sep="\t",quote=FALSE)

  all_res_data[[paste(c,"_lFC",sep="")]]<-res_data$log2FoldChange[match(all_res_data$gene,res_data$gene)]
  all_res_data[[paste(c,"_padj",sep="")]]<-res_data$padj[match(all_res_data$gene,res_data$gene)]
  all_res_data[[paste(c,"_expr",sep="")]]<-ifelse(
    (all_res_data[[paste(c,"_lFC",sep="")]] > 0 & all_res_data[[paste(c,"_padj",sep="")]] < 0.05),"+",ifelse(
      (all_res_data[[paste(c,"_lFC",sep="")]] < 0 & all_res_data[[paste(c,"_padj",sep="")]] < 0.05),"-","0"))
}
contrasts
all_res_data$pat<-paste(all_res_data$H_v_T_expr,all_res_data$T_v_O_expr,all_res_data$O_v_W_expr,sep=",")
```

Annotate each gene by it's pattern of gene expression across fragments.
```{r}
tapply(all_res_data$gene,INDEX=list(all_res_data$pat),length)
all_res_data$annotation<-ifelse(all_res_data$pat=="+,0,0","Testis",
                                ifelse(all_res_data$pat=="0,+,0","Ovary 1",
                                       ifelse(all_res_data$pat=="+,+,0","Ovary 2",
                                              ifelse(all_res_data$pat=="0,0,+","Tail",
                                                     ifelse(all_res_data$pat=="0,0,0","Non-Specific","Other")))))
# Save the results
write.table(all_res_data,
            here::here("rdata/positional_rnaseq","Machtx_all_DESeq2Results_reduced.tab"),
            col.names = TRUE,row.names = FALSE,sep="\t",quote=FALSE)
```

Plot 3D scatter
```{r}
all_res_data<-read.table(here::here("rdata/positional_rnaseq","Machtx_all_DESeq2Results_reduced.tab"),
                         header=TRUE,sep="\t")
normalised_counts<-read.table(here::here("rdata/positional_rnaseq","Machtx_normalised_counts.tab"),sep="\t",header=TRUE)
total<-nrow(count_dat)

# Subset count dat and tpm dat to those that received annotations
normalised_counts<-normalised_counts[which(normalised_counts$gene %in% all_res_data$gene),]

# Compute the mean TPM from across replicates of each fragment
mean_count_dat<-data.frame("gene"=normalised_counts$gene)
mean_count_dat$W_mean<-apply(normalised_counts[,grep("W_",colnames(normalised_counts))],MARGIN = 1,mean)
mean_count_dat$W_expr<-log2(((mean_count_dat$W_mean/sum(mean_count_dat$W_mean)*100)+0.00001))
mean_count_dat$H_mean<-apply(normalised_counts[,grep("H_",colnames(normalised_counts))],MARGIN = 1,mean)
mean_count_dat$H_expr<-log2(((mean_count_dat$H_mean/sum(mean_count_dat$H_mean)*100)+0.00001))
mean_count_dat$T_mean<-apply(normalised_counts[,grep("T_",colnames(normalised_counts))],MARGIN = 1,mean)
mean_count_dat$T_expr<-log2(((mean_count_dat$T_mean/sum(mean_count_dat$T_mean)*100)+0.00001))
mean_count_dat$O_mean<-apply(normalised_counts[,grep("O_",colnames(normalised_counts))],MARGIN = 1,mean)
mean_count_dat$O_expr<-log2(((mean_count_dat$O_mean/sum(mean_count_dat$O_mean)*100)+0.00001))
mean_count_dat$overall_mean<-apply(count_dat[,grep("_",colnames(count_dat))],MARGIN = 1,mean)

mean_count_dat$T_H<-mean_count_dat$T_expr-mean_count_dat$H_expr
mean_count_dat$O_T<-mean_count_dat$O_expr-mean_count_dat$T_expr
mean_count_dat$W_O<-mean_count_dat$W_expr-mean_count_dat$O_expr

mean_count_dat$annotation<-all_res_data$annotation[match(mean_count_dat$gene,all_res_data$gene)]
mean_count_dat$col<-ifelse(mean_count_dat$annotation=="Non-Specific","red",
                           ifelse(mean_count_dat$annotation=="Other","black",
                                  ifelse(mean_count_dat$annotation=="Testis","blue",
                                         ifelse(mean_count_dat$annotation=="Ovary 1","Orange",
                                                ifelse(mean_count_dat$annotation=="Ovary 2","purple",
                                                       ifelse(mean_count_dat$annotation=="Tail","green","grey50"))))))

# Plot 3D scatter plot
setupKnitr()
plot3d(x=mean_count_dat$T_H,
       y=mean_count_dat$O_T,
       z=mean_count_dat$W_O,
       col = mean_count_dat$col,
       xlab = "testis expr - head expr",
       ylab = "ovary expr - testis expr",
       zlab = "whole worm expr - ovary expr",
       type = "p",
       size = 2,
       axis = FALSE
       )
hystrix_scatter <- rglwidget(scene3d(minimal = TRUE))
hystrix_scatter
```